### DISCUSSION TITLE
Handling Overloaded Functions

### DISCUSSION THREAD

**Discussion by arjun-io** [wagmi-dev/abitype#139] [Q&A] [answered]
I'm struggling with using AbiType to handle overloads - it appears that the various input types of are parsed correctly, but the return types are `|`ed together resulting in improper return types

For example, given an ABI: 
```
[
  'function myFunc(address addr) view returns (uint v)',
  'function myFunc(address a, address b) view returns (address v, uint n)'
]
```

the resulting typed output for `contract.read.myFunc(["0xabc123", "0x123abc"])` would be `{ v: bigint | Address }` - note that `n` is missing and the type for `v` is not strict, the expected return type would be `{ v: Address, u: bigint }`

TypeChain would handle this by deduping the function name based on args, so `contract.read["myFunc(address,address)"]` is correctly separated from `contract.read["myFunc(address)"]`

I'm not sure if the type extraction system is expressive enough to handle the overload case? My TypeScript expertise is not quite sufficient to be able to understand how to accomplish something like this.

**Comment by Raiden1411** [comment]
We currently are aware that this is a problem on overloads. Before wagmi v1 the workaround for this was similar. [Reference](https://github.com/wagmi-dev/wagmi/blob/38f03dabde9c5aebfbe97565e18998c52a889bbc/packages/core/src/actions/contracts/getContract.ts#L189)

However with the introduction of viem this no longer applies. The current idea to solve this problem is to narrow the return type based on the provided arguments on those overload functions. So in your example if you just pass the one address as the argument then it would be able to infer the correct return type based on it.

This is currently on the "whiteboard" and should be something that gets solved in the near future.

**Comment by Raiden1411** [comment, marked as answer]
Example implementation merged in #149

**Reply by arjun-io** [reply]
Awesome - will further work need to be done in Viem to support this for their contract types or should I be able to simply add a `resolution` here?

**Reply by tmm** [reply]
Some work needs to be done to bring this over to viem.

**Reply by tmm** [reply]
Following up that Viem supports overloads since v2.
